### PR TITLE
add GatherSliceToSplitFusion and Unittest

### PR DESCRIPTION
**Pull request by rui-ren**
### Multi Query Attention Optimization

in multi-query attention
```
batch_size, seq_length, three_times_hidden_size = fused_qkv.shape
fused_qkv = fused_qkv.view(batch_size, seq_length, self.num_heads + 2, self.head_dim)
return fused_qkv[..., :-2, :], fused_qkv[..., [-2], :], fused_qkv[..., [-1], :]
```
which can be optimized to 
```
batch_size, seq_length, three_times_hidden_size = fused_qkv.shape
fused_qkv = fused_qkv.view(batch_size, seq_length, self.num_heads + 2, self.head_dim)
(query, key, value) = fused_qkv.split([self.num_heads, 1, 1], dim=2)
return query, key, value
```

this optimization can be validated from nsight profiling and perf benchmarking.
   
image

As such, This PR is to Optimize the `Gather/Gather/Slice` Ops to `Split` Kernel.

### Optimization Target

As 2 `Gather` and 1 `Slice` Kernels are time consuming for backward prop, it would be efficient to use 1 `Split` Kernel  

###  Example

- Before Fusion
image
 
- After Fusion
image

### Perf Gain
After the optimization, there will have **~7%** perf gain. 

> The `Transpose` Kernel can be fused too, will update it in next PR.  However, after testing Transponse Ops fusion on Falcon model, there is no perf gain. Will not create a new PR.

### CODE REVIEW

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 1]
```diff
@@ -7642,5 +7643,87 @@ TEST_F(GraphTransformationTests, GatherToSliceFusion) {
...
+TEST_F(GraphTransformationTests, GatherSliceToSplitFusion) {
+  {
+    auto build_test_case = [&](ModelTestBuilder& builder) {
+      auto* data_arg = builder.MakeInput<float>({{54}});
+      auto* reshape_arg = builder.MakeInput<int64_t>({{4}});
+      auto* reshape_out = builder.MakeIntermediate<float>({{2, 512, 73, 64}});
+      builder.AddNode("Reshape", {data_arg, reshape_arg}, {reshape_out});
+
+      // Create Shape-0 Ops
+      auto* shape_output_0 = builder.MakeOutput();
```
Comment (zhijxu-MS) [inline onnxruntime/test/optimizer/graph_transform_test.cc:7655]: why we need shape here? just 1reshape+2gather+1slice is enough to test your optimizer

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 2]
```diff
@@ -7642,5 +7643,87 @@ TEST_F(GraphTransformationTests, GatherToSliceFusion) {
   }
 }
 
+TEST_F(GraphTransformationTests, GatherSliceToSplitFusion) {
```
Comment (zhijxu-MS) [inline onnxruntime/test/optimizer/graph_transform_test.cc:7646]: add a test to cover different dims?

also, add a test to cover the case when your optimizer can't optimize it, for example gather in diff dims

File: onnxruntime/test/optimizer/graph_transform_test.cc (hunk 1 above)
Comment (rui-ren) [reply onnxruntime/test/optimizer/graph_transform_test.cc]: Yes, deleted the shape ops, thank you!

File: onnxruntime/core/optimizer/gather_slice_fusion.cc [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    for (auto node_index : node_topology_list) {
+        auto* p_node = graph.GetNode(node_index);
+
+        if (p_node == nullptr) continue;
+
+        Node& node = *p_node;
+
+        ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+        // Currently only catch after Reshape ops, optimize in the future
```
[the remark this answers is not in the document]
Comment (rui-ren) [reply onnxruntime/core/optimizer/gather_slice_fusion.cc:179]: Thanks!

File: onnxruntime/core/optimizer/gather_slice_fusion.cc [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+    return true;
+}
+
+
+bool GatherSliceToSplitFusion::IsSupportedSlice(const Graph& graph, const Node& node,
+                                                InlinedVector<int64_t>& starts,
+                                                InlinedVector<int64_t>& ends,
+                                                InlinedVector<int64_t>& axes,
+                                                InlinedVector<int64_t>& steps) const {
+    // check the version of Slice ops
```
[the remark this answers is not in the document]
Comment (rui-ren) [reply onnxruntime/core/optimizer/gather_slice_fusion.cc:52]: Thanks!

File: onnxruntime/core/optimizer/gather_slice_fusion.cc [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+
+
+namespace onnxruntime {
+
+bool GatherSliceToSplitFusion::IsSupportedGather(const Graph& graph, const Node& node, int64_t& index,
+                                                    int64_t& axis, int64_t& indices_n_dims) const {
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Gather", {1, 11, 13}) ||
+        !graph_utils::IsSupportedProvider(node, GetCompatibleExecutionProviders())) {
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (rui-ren) [reply onnxruntime/core/optimizer/gather_slice_fusion.cc:18]: Thanks!

File: onnxruntime/test/optimizer/graph_transform_test_builder.cc [hunk 6]
```diff
@@ -15,7 +15,7 @@
 #include "test/util/include/inference_session_wrapper.h"
 
 // enable to dump model for debugging
-#define SAVE_TEST_GRAPH 0
+#define SAVE_TEST_GRAPH 1
```
Comment (hanbitmyths) [inline onnxruntime/test/optimizer/graph_transform_test_builder.cc:18]: Revert back to 0, so it won't be checked in.

Comment (rui-ren) [reply onnxruntime/test/optimizer/graph_transform_test_builder.cc]: Updated, thanks!

File: onnxruntime/test/optimizer/graph_transform_test.cc (hunk 2 above)
Comment (rui-ren) [reply onnxruntime/test/optimizer/graph_transform_test.cc:7646]: Updated, thanks!!, please check the unitests.

File: onnxruntime/core/optimizer/gather_slice_fusion.cc [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+        InlinedVector<int64_t> split_value{{slice_dim, 1, 1}};
+        split_initializer_proto.set_raw_data(split_value.data(), split_value.size() * sizeof(int64_t));
+        NodeArg* split_arg = &graph_utils::AddInitializer(graph, split_initializer_proto);
+
+        Node& split_node =
+            graph.AddNode(graph.GenerateNodeName("Split"), "Split", "Split for fused Gather-Slice fusion",
+                                                {graph.GetNodeArg(node_arg->Name()), split_arg}, split_outputs);
+
+        split_node.AddAttribute("axis", split_axis);
+        // to do here
```
[the remark this answers is not in the document]
Comment (rui-ren) [reply onnxruntime/core/optimizer/gather_slice_fusion.cc]: deleted, thanks!

**Review by zhijxu-MS — APPROVED** [review]

**Review by askhade — APPROVED** [review]
LGTM

**Pull request merged**